Add jose dependency to support ES***, PS***, and RS*** algorithms.

potatosalad opened the pull request:
Following up on the comment I made about a week ago in #13, I integrated version 1.0.1 of the [jose](https://github.com/potatosalad/erlang-jose) project as the backend for signing and verifying.

This adds support for the following signing algorithms (JWS `"alg"`):

* ES256
* ES384
* ES512
* PS256
* PS384
* PS512
* RS256
* RS384
* RS512

As well as supporting the following key types (JWK `"kty"`):

* EC
* RSA
* oct

If a map or JWK isn't specified for the `secret_key`, it's assumed to be a `oct` symmetric key (so no change in the default behavior already present in joken).

While not present in this pull request, there's also the potential for integrating the `sph` and `b64` fields from the [JWS Signing Input Options](https://tools.ietf.org/html/draft-ietf-jose-jws-signing-input-options-00) working draft.  That would require more than just the simple `alg` field returned from the configuration, however.

You can convert private/public key PEM files you have to the JWK format using (see the P-256 EC key below as an example):

```elixir
jwk = JOSE.JWK.from_pem_file("rsa.pem")
{_, rsa_map} = JOSE.JWK.to_map(jwk)
```

Here is an example of ES256:

```elixir
defmodule My.Config.Module do
  @behaviour Joken.Config

  def secret_key() do
    %{"crv" => "P-256", "d" => "aJhYDBNS-5yrH97PAExzWNLlJGqJwFGZmv7iJvdG4p0",
      "kty" => "EC", "x" => "LksdLpZN3ijcn_TBfRK-_tgmvws0c5_V5k0bg14RLhU",
      "y" => "ukc-JOEAWhW664SY5Q29xHlAVEDlrQwYF3-vQ_cdi1s"}
  end
  
  def algorithm(), do: :ES256

  def encode(map), do: :jsx.encode(map)

  def decode(binary) do
    :jsx.decode(binary)
    |> Enum.map(fn({key, value})-> {String.to_atom(key), value} end)
  end

  def claim(:exp, payload) do
    Joken.Helpers.get_current_time() + 300
  end

  def claim(_, _), do: nil

  def validate_claim(:exp, payload, options) do
    Joken.Helpers.validate_time_claim(payload, "exp", "Token expired", fn(expires_at, now) -> expires_at > now end)
  end

  def validate_claim(_, _, _), do: :ok
end
```

```elixir
{:ok, token} = Joken.Token.encode(My.Config.Module, %{username: "johndoe"})
```

```elixir
{:ok, decoded_payload} = Joken.Token.decode(My.Config.Module, jwt)
```

cs-victor-nascimento commented:
This is great work! I just have some questions (by no means I am a crypto expert... please bare with me):

- Why have you implemented the whole AES algorithm straight in Erlang? Couldn't we just have used [crypto](http://www.erlang.org/doc/man/crypto.html) as with the other implementations? I believe it will be more performant because it is a NIF (if I am not mistaken) and even if it is not a NIF it is certainly more revised code than a third party implementation. 
- Your are depending on base64url which I believe is not Apache 2.0. I think we could have a base64url helper in JOSE to lower our dependencies. Besides the license, we could provide a more thorough test suite for it.
- I haven't read all code in erlang-jose yet but one breaking change this PR brings is that Joken excluded the none algorithm on purpose. This brings it back and I guess it would be better to leave it out (though it is in the RFC). See #32.
- I guess we could go all in with this PR. So, changing the algorithm callback to something that supports JWS all the way would be nice to be here.

cs-victor-nascimento commented:
Also, tests are failing only in Erlang 17.5 in Travis.

potatosalad commented:
@cs-victor-nascimento OTP 18 and newer are the only supported versions of Erlang (as of right now) due to 3 main reasons:

1. `aes_ecb` was added in OTP 18
2. `aes_gcm` was added in OTP 18
3. EC keys changed from lists and tuples to straight binaries in OTP 18 (see [public_key: Remove legacy switch compact_bit_string](https://github.com/erlang/otp/commit/3bf1096068c3a123a23e6a1499152976a2da131e))

Now to address your questions:

> Why have you implemented the whole AES algorithm straight in Erlang? Couldn't we just have used crypto as with the other implementations? I believe it will be more performant because it is a NIF (if I am not mistaken) and even if it is not a NIF it is certainly more revised code than a third party implementation.

I completely agree that `crypto` should be used over a pure Erlang AES implementation.  However, `crypto` only supports `aes_cbc128` and `aes_cbc256` (so no `aes_cbc192`) and `aes_ecb` fails if you try to use a 192-bit key.  The pure Erlang AES encryption is only used if a 192-bit key is detected in order to provide support for all of the algorithms mentioned in the [JSON Web Algorithms (JWA) RFC](https://tools.ietf.org/html/rfc7518).

In relation to this, if we are to support 17.5 and below, we would also need to provide fallbacks for 128-bit, 192-bit, and 256-bit encryption and decryption for the `aes_ecb` and `aes_gcm` algorithms.  Fallbacks for the EC key format change would also need to be integrated.

> Your are depending on base64url which I believe is not Apache 2.0. I think we could have a base64url helper in JOSE to lower our dependencies. Besides the license, we could provide a more thorough test suite for it.

[base64url](https://github.com/dvv/base64url) uses the MIT license, which as far as I know, has no conflict with the Apache 2.0 license.  It looked like people were already contributing to the base64url project, so I figured that would be a safer bet for future fixes than trying to include a static copy in the project itself.

> I haven't read all code in erlang-jose yet but one breaking change this PR brings is that Joken excluded the none algorithm on purpose. This brings it back and I guess it would be better to leave it out (though it is in the RFC). See #32.

The `none` algorithm type now has a different behavior than the one that was present in Joken previously.  It fully expects the signature to be an empty binary.  If it is not an empty binary or the `"alg"` is not `"none"`, then it fails validation.  See [JSON Web Signature (JWS) - A.5. Example Unsecured JWS](https://tools.ietf.org/html/rfc7515#appendix-A.5) and [jose_jws_alg_none.erl](https://github.com/potatosalad/erlang-jose/blob/master/src/jose_jws_alg_none.erl) for more information.

> I guess we could go all in with this PR. So, changing the algorithm callback to something that supports JWS all the way would be nice to be here.

With the current restriction on OTP 18 present (until fallbacks can be fully implemented for 17.5), it might make sense to have [jose](https://github.com/potatosalad/erlang-jose) as an optional dependency that can be used by Joken if detected as present, but fallback to `crypto:hmac/3` if not present.

potatosalad commented:
@cs-victor-nascimento I misunderstood your last comment when I first read it, so I wanted to add onto my response:

> I guess we could go all in with this PR. So, changing the algorithm callback to something that supports JWS all the way would be nice to be here.

We're kind of doing that already with the `secret_key` function, so it might make sense to do that with the `algorithm` function, too.  They could potentially be renamed to reflect their actual intent, for example:

```elixir
def json_web_key() do
  %{"kty" => "oct",
    "k" => "bXkgc2VjcmV0IGtleQ"}
end

def json_web_signature() do
  %{"alg" => "HS256",
    "b64" => false,
    "cty" => "JWT",
    "sph" => true}
end
```

Or if you have a RSA PEM file:

```elixir
def json_web_key() do
  JOSE.JWK.from_pem_file("rsa.pem")
end

def json_web_signature() do
  %{"alg" => "RS256",
    "b64" => false,
    "cty" => "JWT",
    "sph" => true}
end
```

cs-victor-nascimento commented:
Hmmm... You are right! Did not know about these key lengths had issues in otp. 

Elixir 1.0.x is meant to be compatible with Erlang 17+. Elixir 1.1 will be compatible with Erlang 18+. So, IMHO it is better to raise when the Erlang version does not support an algorithm. My reasoning is to switch back to crypto implementation of AES and document that it won't work with keys of 192 bit size nor with those other modes of operation for this block cipher. Even if that means we are not full compliant with the JWE, JWS and JWT RFCs. 

I could not see the MIT License in base64url there! My bad! Totally agree with you now.

This work is wonderful and I think your alternative for secret_key is awesome too. 

@bryanjos what do you think about it all?

bryanjos commented:
@potatosalad great work! I'll admit I don't know much about cryptography, so thanks for clarifying some of the changes. 

@cs-victor-nascimento @potatosalad I agree Erlang 17.5 support is important (I'm still using it until homebrew updates to 18). We should probably not support any of the algorithms that won't work there yet. Or if possible somehow make those optional.

As long as the algorithm is still being passed into the signature verification function instead of the one used on the token we should be ok. The thing with `none` is I think there was some confusion between its purpose and the suggested security fix for JWT verification that caused some to think it is automatically unsafe. I'm fine with including it if it makes sense, but if it'll cause too much confusion, or if it's something that's not used much in practice, it may be good to not use it.

potatosalad commented:
@bryanjos thanks!

@cs-victor-nascimento @bryanjos I changed the way jose detects which ciphers are available allows for fallback to the slower pure Erlang implementation (required adding an AES GCM implementation).  I also have it detecting the different formats of EC keys between OTP 17.5 and OTP 18.  Those few changes allow it to run everything on both version of OTP.

In addition to supporting JWS and JWK, it might make sense to add optional support for JWE for encrypted and nested encrypted JWT (see [A.1 and A.2 of JWT RFC](https://tools.ietf.org/html/rfc7519#appendix-A.1)):

```elixir
def json_web_encryption() do
  %{"alg" => "dir",
    "cty" => "JWT",
    "enc" => "A128CBC-HS256"}
end

def json_web_key() do
  %{"kty" => "oct",
    "k" => "no6iPLY8Io8xrZJotAl_FW1qoR3LKfFLrM_xls4_CtI"}
end

def json_web_signature() do
  %{"alg" => "HS256",
    "b64" => false,
    "cty" => "JWT",
    "sph" => true}
end
```

Then have something like this:

```elixir
{:ok, encrypted} = Joken.encrypt(%{test: true})
# {:ok, "eyJhbGciOiJkaXIiLCJjdHkiOiJKV1QiLCJlbmMiOiJBMTI4Q0JDLUhTMjU2In0..kQ9FuxEhPX_Ji15Nb0SF_g.wBmocAUttSJc2TGnk6efMQ.HN2odour_F6-P7GXGB75uw"}

Joken.decrypt(encrypted)
# {:ok, %{test: true}}
```

All of this might be more appropriate to discuss as a separate pull request.

cs-victor-nascimento commented:
@potatosalad  I see you are saving the supported crypto algorithms in an ETS table. Nice!

This still leaves a custom AES implementation in some operation modes in pure Erlang. Again, I am no cryptographer and your code seems quite nice. More than that: I've seen you contributing to crypto itself which gives the implementation more credit. But I guess that having at least a big warning or a mandatory config change for making the fallback algorithms opt-in and not opt-out would definitely benefit users. 

Other than that I think this is what is needed for 1.0.

bryanjos commented:
@potatosalad I see for erlang-jose you have some notes on the custom implementations so maybe just linking to there in the README after the supported algorithms would suffice. What do you think @cs-victor-nascimento ?

Other than that, :+1:

cs-victor-nascimento commented:
I think we should merge it now that it passes on both Erlang versions. We are discussing some erlang-jose issues now. This discussion is better to be had over there.
:+1:

potatosalad commented:
@bryanjos For the signature algorithms, PS256, PS384, and PS512 are the only ones specifically related to this pull request that are not fully native, so I added a note and a link about that to the README.

@cs-victor-nascimento I've added issues potatosalad/erlang-jose#1 and potatosalad/erlang-jose#2 as a continuation of some of the ideas we've discussed here for algorithm fallback.

bryanjos commented:
@potatosalad merged! Thanks for doing this!

cs-victor-nascimento commented:
Great work and thanks for your patience!
